src/protobuf.zig

```diff
@@ -980,6 +1345,159 @@ pub fn MessageMixins(comptime Self: type) type {
         pub fn dupe(self: Self, allocator: Allocator) !Self {
             return pb_dupe(Self, self, allocator);
         }
+        pub fn json_decode(
+            input: []const u8,
+            options: json.ParseOptions,
+            allocator: Allocator,
+        ) !std.json.Parsed(Self) {
+            return pb_json_decode(Self, input, options, allocator);
+        }
+        pub fn json_encode(
+            self: Self,
+            options: json.StringifyOptions,
+            allocator: Allocator,
+        ) ![]const u8 {
+            return pb_json_encode(self, options, allocator);
+        }
+
+        // This method is used by std.json
+        // internally for deserialization. DO NOT RENAME!
+        pub fn jsonParse(
+            allocator: Allocator,
+            source: anytype,
+            options: json.ParseOptions,
+        ) !Self {
+            if (.object_begin != try source.next()) {
+                return error.UnexpectedToken;
+            }
+
+            // Mainly taken from 0.13.0's source code
+            var result: Self = undefined;
+            const structInfo = @typeInfo(Self).Struct;
+            var fields_seen = [_]bool{false} ** structInfo.fields.len;
+
+            while (true) {
+                var name_token: ?json.Token = try source.nextAllocMax(
+                    allocator,
+                    .alloc_if_needed,
+                    options.max_value_len.?,
+                );
+                const field_name = switch (name_token.?) {
+                    inline .string, .allocated_string => |slice| slice,
+                    .object_end => { // No more fields.
+                        break;
+                    },
+                    else => {
+                        return error.UnexpectedToken;
+                    },
+                };
+
+                inline for (structInfo.fields, 0..) |field, i| {
+                    if (field.is_comptime) {
+                        @compileError("comptime fields are not supported: " ++ @typeName(Self) ++ "." ++ field.name);
+                    }
+
+                    const yes1 = std.mem.eql(u8, field.name, field_name);
+                    var yes2: bool = undefined;
+
+                    const camel_case_name = try to_camel_case(field.name, allocator);
+                    if (field.name.ptr == camel_case_name.ptr) {
+                        // If there was no case conversion than the same
+                        // array will be returned (there's no need
+                        // for deallocation in this case)
+                        yes2 = false;
+                    } else {
+                        yes2 = std.mem.eql(u8, camel_case_name, field_name);
+                        allocator.free(camel_case_name);
+                    }
+
+                    if (yes1 and yes2) {
+                        return error.FieldNameCollision;
```

Comment:
This results in:
```
/usr/lib/zig/std/json/static.zig:322:35: note: 'error.FieldNameCollision' not a member of destination error set
/usr/lib/zig/std/json/static.zig:211:32: note: function return type declared here
) ParseError(@TypeOf(source.*))!T {
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
referenced by:
```



I think you need:
```suggestion
                        return error.DuplicateField;
```
which is in the json error set.

Reply:
That's the type of confusion I talked about. IMO DuplicateField throws when JSON string contains two entries of one key, but that's the opposite case when one dictionary entry match with two struct fields. It's bad naming anyway and user should avoid it but it's still valid code and because of this specific error he might spend more time finding what's wrong and why his code is not working. That's why I thought to use panic - if you want to work with **bad** structure without JSON conversion everything will be fine, but as soon as you'll try to (de)serialize it via JSON there will a panic "Boy you'll shoot yourself in the leg with this". Is it good to not let user generate structures from these **bad** message definitions?

Maybe it's fine and I'm overreacting :) I think it will be the rare case anyway that someone would choose different name style for message fields using the same words but different *delimiters*.